### DISCUSSION TITLE
Staticcheck keeps showing false warnings pertaining to code in files of package main. How to fix?

### DISCUSSION THREAD

**Discussion by squint2** [golang/vscode-go#3169] [Help]
Types and functions that exist in files of package main are not being recognized in main.go, so they flag as undefined warnings by the staticchecker(?!).

The app runs without issues, but it's annoying to see so many warnings that are false. One solution is place all code in the main.go file, but that is not great either.

Is there a way to fix this problem?  I like having the staticchecker, but also want it to work properly.

Sam

**Comment by zsolt-p** [comment]
+1

**Comment by toelt** [comment]
+1

**Comment by 0x48core** [comment]
This is a known issue with how `gopls` (and staticcheck via gopls) handles multi-file `package main` projects. The root cause is usually that **gopls isn't treating all your files as part of the same package** — often because the project isn't recognized as a proper Go module or the workspace root isn't set correctly.

Here are a few things to check:

**1. Make sure you have a `go.mod` file at your project root**

Run this if you haven't already:
```bash
go mod init your-module-name
```
Without `go.mod`, gopls can't properly resolve multi-file `package main` packages.

**2. Open VS Code at the correct root folder**

Open the folder that contains `go.mod` directly — not a parent folder. gopls determines package scope from the workspace root.

**3. Check your `gopls` workspace settings**

In `.vscode/settings.json`, make sure you're not restricting the analysis scope:
```json
{
  "gopls": {
    "analyses": {
      "staticcheck": true
    },
    "directoryFilters": []
  }
}
```

**4. If using a multi-folder workspace (`.code-workspace`)**

Each folder entry must point to the module root individually, otherwise gopls can lose track of sibling files in `package main`.

---

After making changes, run **"Go: Restart Language Server"** from the VS Code command palette (`Ctrl+Shift+P`) to apply them.

Hope this helps! Let me know which step resolves it. 🙏
